tests/drivers/i2c/i2c_emul/src/test_forwarding.cpp

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright 2024 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <functional>
+
+#define CUSTOM_FFF_FUNCTION_TEMPLATE(RETURN, FUNCNAME, ...)                                        \
+	std::function<RETURN(__VA_ARGS__)> FUNCNAME
+
+#include "emulated_target.h"
+
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/fff.h>
+#include <zephyr/ztest.h>
+
+DEFINE_FFF_GLOBALS;
+
+#define CONTROLLER_LABEL DT_NODELABEL(i2c0)
+#define TARGET_LABEL     DT_NODELABEL(i2c1)
+
+namespace
+{
+/* Get the devicetree constants */
+constexpr const struct device *controller = DEVICE_DT_GET(CONTROLLER_LABEL);
+constexpr const struct device *target = DEVICE_DT_GET(TARGET_LABEL);
+constexpr const uint16_t forward_address = DT_PHA_BY_IDX(CONTROLLER_LABEL, forwards, 0, addr);
+
+BUILD_ASSERT(DEVICE_DT_GET(DT_PHANDLE_BY_IDX(CONTROLLER_LABEL, forwards, 0)) == target,
+	     "Forwarding misconfigured");
+
+/* Setup the fake functions for the I2C target. */
+FAKE_VALUE_FUNC(int, target_read_requested, struct i2c_target_config *, uint8_t *);
+FAKE_VALUE_FUNC(int, target_read_processed, struct i2c_target_config *, uint8_t *);
+FAKE_VALUE_FUNC(int, target_write_requested, struct i2c_target_config *);
+FAKE_VALUE_FUNC(int, target_write_received, struct i2c_target_config *, uint8_t);
+FAKE_VALUE_FUNC(int, target_stop, struct i2c_target_config *);
+
+static void *i2c_emul_forwarding_setup(void)
+{
+	// Register the target
+	zassert_ok(i2c_target_register(target, &emulated_target_config));
+
+	// Setup the fake callbacks
+	emulated_callbacks.read_requested = target_read_requested;
+	emulated_callbacks.read_processed = target_read_processed;
+	emulated_callbacks.write_requested = target_write_requested;
+	emulated_callbacks.write_received = target_write_received;
+	emulated_callbacks.stop = target_stop;
+
+	return NULL;
+}
+
+static void i2c_emul_forwarding_before(void *fixture)
+{
+	ARG_UNUSED(fixture);
+
+	// Reset all fakes
+	RESET_FAKE(target_read_requested);
+	RESET_FAKE(target_read_processed);
+	RESET_FAKE(target_write_requested);
+	RESET_FAKE(target_write_received);
+	RESET_FAKE(target_stop);
+	FFF_RESET_HISTORY();
+}
+
+static void i2c_emul_forwarding_teardown(void *fixture)
+{
+	ARG_UNUSED(fixture);
+
+	// Unregister the I2C target callbacks
+	zassert_ok(i2c_target_unregister(target, &emulated_target_config));
+}
+
+ZTEST_SUITE(i2c_emul_forwarding, NULL, i2c_emul_forwarding_setup, i2c_emul_forwarding_before, NULL,
+	    i2c_emul_forwarding_teardown);
+
+ZTEST(i2c_emul_forwarding, test_write_is_forwarded)
```

Comment:
I think a `test_transfer_is_forwarded` would also be useful to validate a multi-stage transfer.

Reply:
Done, man, lambdas for the win on this.